common/multistep_runner.go

```diff
@@ -0,0 +1,163 @@
+// for -debug and -on-error command line arguments.
+func NewRunner(steps []multistep.Step, config PackerConfig, ui packer.Ui) multistep.Runner {
+	runner, _ := newRunner(steps, config, ui)
+	return runner
+}
+
+// NewRunnerWithPauseFn returns a multistep.Runner that runs steps augmented
+// with support for -debug and -on-error command line arguments.  With -debug it
+// puts the multistep.DebugPauseFn that will pause execution between steps into
+// the state under the key "pauseFn".
+func NewRunnerWithPauseFn(steps []multistep.Step, config PackerConfig, ui packer.Ui, state multistep.StateBag) multistep.Runner {
+	runner, pauseFn := newRunner(steps, config, ui)
+	if pauseFn != nil {
+		state.Put("pauseFn", pauseFn)
+	}
+	return runner
+}
+
+type packerRunner struct {
+	runner  multistep.Runner
+	pauseFn interface{}
+}
+
+func (r *packerRunner) Run(s multistep.StateBag) { r.runner.Run(s) }
+func (r *packerRunner) Cancel()                  { r.runner.Cancel() }
+func (r *packerRunner) PauseFn() interface{}     { return r.pauseFn }
+
+func typeName(i interface{}) string {
+	return reflect.Indirect(reflect.ValueOf(i)).Type().Name()
+}
+
+type abortStep struct {
+	step multistep.Step
+	ui   packer.Ui
+}
+
+func (s abortStep) Run(state multistep.StateBag) multistep.StepAction {
+	return s.step.Run(state)
+}
+
+func (s abortStep) Cleanup(state multistep.StateBag) {
+	if _, ok := state.GetOk(multistep.StateCancelled); ok {
+		s.ui.Error("Interrupted, aborting...")
+		os.Exit(1)
+	}
+	if _, ok := state.GetOk(multistep.StateHalted); ok {
+		s.ui.Error(fmt.Sprintf("Step %q failed, aborting...", typeName(s.step)))
+		os.Exit(1)
+	}
+	s.step.Cleanup(state)
+}
+
+type askStep struct {
+	step multistep.Step
+	ui   packer.Ui
+}
+
+func (s askStep) Run(state multistep.StateBag) (action multistep.StepAction) {
+	for {
+		action = s.step.Run(state)
+
+		if action != multistep.ActionHalt {
+			return
+		}
+
+		switch ask(s.ui, typeName(s.step), state) {
+		case askCleanup:
+			return
+		case askAbort:
+			os.Exit(1)
+		case askRetry:
+			continue
+		}
+	}
+}
+
+func (s askStep) Cleanup(state multistep.StateBag) {
+	s.step.Cleanup(state)
+}
+
+type askResponse int
+
+const (
+	askCleanup askResponse = iota
+	askAbort
+	askRetry
+)
+
+func ask(ui packer.Ui, name string, state multistep.StateBag) askResponse {
+	ui.Say(fmt.Sprintf("Step %q failed", name))
+
+	result := make(chan askResponse)
+	go func() {
+		result <- askPrompt(ui)
+	}()
+
+	for {
+		select {
+		case response := <-result:
+			return response
+		case <-time.After(100 * time.Millisecond):
+			if _, ok := state.GetOk(multistep.StateCancelled); ok {
+				return askCleanup
+			}
+		}
+	}
+}
+
+func askPrompt(ui packer.Ui) askResponse {
+	for {
+		line, err := ui.Ask("[C]lean up and exit, [A]bort without cleanup, or [R]etry step (build may fail even if retry succeeds)? [car]")
+		if err != nil {
+			log.Printf("Error asking for input: %s", err)
+		}
+
+		switch {
+		case len(line) == 0 || line[0] == 'c':
```

Comment:
I would make this a case-insensitive comparison, and remove the redundant "[car]" from the message above.